### DISCUSSION TITLE
Interesting on developing complete web hosting plugin.

### DISCUSSION THREAD

**Discussion by vireakvibol** [ajenti/ajenti#1397] [General]
Hello! I'm planning to development a complete **web hosting plugin** to manage share hosting environment. I want to do this just for learning and exploring purpose.

I'm interesting the [ajenti-3-dev](https://github.com/ajenti/ajenti/tree/ajenti-3-dev) branch as it and redesign on plugin development and angular2 integration as i'm currently working on angular based as front-end project.

Is **ajenti** design to capable on managing many user with different role like admin, reseller, end-user like other panel dose? Or any suggestion for me to getting start on this experience?

**Comment by daniel-schulz** [comment]
Hi @vireakvibol,

Welcome and thank you that you're interested in supporting Ajenti!

Short information regarding [ajenti-3-dev branch](https://github.com/ajenti/ajenti/tree/ajenti-3-dev) in advance:
We are still working on basic features and some architecture details for Ajenti 3. Currently, there are some extra steps necessary, but we will improve the developer experience as fast as possible and continuously. 

Regarding your "group" requirement: As I know, Ajenti is basically using POSIX authentication, but can also use directory services, like Samba, to organise groups and rights via extra plugins. I guess @kiarn can give you deeper insights, as he is our Ajenti master and knows almost everything about all features and plugins ;-)

Furthermore I know, there are plans to improve many of the current authentication and authorisation capabilities like implementing OAuth, JWT, ... 

Just reach me out, if you need any help to setup your development environment or have questions about the new angular2-based architecture.

Best,
Daniel

**Comment by kiarn** [comment]
Hello @vireakvibol,

I'm glad to see your message here, welcome !

First, it would be good to setup a dev environment. We have started a documentation in order to explain some steps for development, but maybe there's some points to clarify : https://docs.ajenti.org/en/ajenti-3-dev/dev/setup-devenv-extension-plugins.html#

> Is **ajenti** design to capable on managing many user with different role like admin, reseller, end-user like other panel dose? Or any suggestion for me to getting start on this experience?

Yes, but it depends on what your want to achieve, and how you want to authenticate your users.
Regarding your project, I suppose the users data will be stored in a database, or am I wrong ?

The modular structure of `Ajenti` let you use your own [AuthenticationProvider](https://github.com/ajenti/ajenti/blob/4469f66f8e87159f177618c2b4c7b12a2aa44973/ajenti-core/aj/auth.py#L72).
I currently know 3 differents `AuthenticatorProviders` : 

  * OS, based on PAM, integrated in `Ajenti` : [OSAuthProvider](https://github.com/ajenti/ajenti/blob/4469f66f8e87159f177618c2b4c7b12a2aa44973/ajenti-core/aj/auth.py#L108)
  * via users file, avalaible as plugin in `Ajenti` : [UsersAuthProvider](https://github.com/ajenti/ajenti/blob/4469f66f8e87159f177618c2b4c7b12a2aa44973/plugins/auth_users/api.py#L15)
  * LDAP auth in particular environment, developed in [linuxmuster.net](https://github.com/linuxmuster) project : [LdapAuthProvider](https://github.com/linuxmuster/linuxmuster-webui7/blob/b7a5416ebe9649dceea045ddfa7014968c2769ce/usr/lib/linuxmuster-webui/plugins/lmn_auth/api.py#L28)

It's possible to make your own Authenticator and to define yourself the roles of the users. The base AuthenticationProvider object supports the following methods : 
  * `authenticate` : method to define in order to check the password,
  * `authorize` : define if a user has permission to access a particular resource,
  * `get_isolation_uid` and `get_isolation_gid` : each authenticated user access `Ajenti` via an isolated worker in order to avoid increase security, and these functions just define the uid and gid for the worker to run with,
  * `get_profile` : get user's profile,
  * `check_email` and `check_password_complexity` : functions for email reset functionality,
  * `update_password` : pretty clear,
  * `prepare_environment` : actions to perform in order to prepare user's environment after login,
  * `signout` : actions to perform in order to cleanup user's environment before closing the worker at logout.

I'm not really a Ajenti master, that's more Eugeny, but that's what I learned during the developpment :)

Please let me know you need more details to achieve your project.

Cheers

Arnaud

**Comment by vireakvibol** [comment]
Dear @daniel-schulz and @kiarn Thank for your warm welcome. I will start setting up a development environment first and let see how its going. After that i will approach my college to join and giving me a hand on the development process. 

**Regarding on my question** I saw on the docs on **ajenti-3-dev** to start the development need to install **ajenti/core** and the command is
`curl https://raw.githubusercontent.com/ajenti/ajenti/master/scripts/install.sh | sudo bash -s -` still point to master branch.
Does it suppose to
`curl https://raw.githubusercontent.com/ajenti/ajenti/ajenti-3-dev/scripts/install.sh | sudo bash -s -` ?

Thank you
Vibol

**Comment by kiarn** [comment]
There's no difference between both files. This script will install all package dependencies to use `Ajenti`, that's why it is required.
The only main structure difference between both branch is the directory `plugins-new` which actually contains 3 plugins developed under Angular 13. The directory `plugins` still contains the actual plugins.

Arnaud

**Reply by vireakvibol** [reply]
Thank you for clarify.

**Reply by vireakvibol** [reply]
Currently running
`ajenti-dev-multitool --build-frontend`

> option --build-frontend not recognized

Do i miss something?

**Reply by vireakvibol** [reply]
Already rise the issue to https://github.com/ajenti/ajenti-dev-multitool/issues/17 let discuss it there.

**Comment by kiarn** [comment]
You didn't miss anything : I didn't took the time to merge the new version of `ajenti-dev-multitool`, that's all, sorry for that.
I do not have the time to do it before thursday, so I just copied it to my own fork, you can find the right script here : https://github.com/kiarn/ajenti-dev-multitool/blob/ajenti-3-dev/ajenti-dev-multitool

Arnaud
